### ISSUE TITLE
Photos (HEIC/JPG) are corrupted when uploading via iOS app

### ISSUE THREAD

**Issue by obale** [immich-app/immich#14587]
### The bug

Photos (HEIC) uploaded via the Immich iOS app are showing up as corrupted in the Immich web interface and as (corrupted) replicated image on mobile. The corrupted photos are visualized flawed and are showing green (anod other colored) artifacts. These artifacts are not always in the same location when deleting and re-uploading the photo, at least as far as I can tell.

In version v1.121.0 and before I didn't experience this issue. Downgrading the server to v1.121.0 didn't solve the issue, so it may be related to the iOS version or a combination of both (server and mobile).

Was searching for a related issue, but couldn't find any.

### The OS that Immich Server is running on

Ubuntu 24.04.1 LTS

### Version of Immich Server

v1.122.1

### Version of Immich Mobile App

v1.122.0

### Platform with the issue

- [X] Server
- [X] Web
- [X] Mobile

### Your docker-compose.yml content

```YAML
#                        
# WARNING: Make sure to use the docker-compose.yml of the current release:
#              
# https://github.com/immich-app/immich/releases/latest/download/docker-compose.yml
#                                           
# The compose file on main may not be compatible with the latest release.
#                            
                                                                       
name: immich                 
                                                                       
services:          
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    volumes:  
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /etc/localtime:/etc/localtime:ro
      - /mnt/nextcloud:/mnt/nextcloud:ro
    env_file:
      - .env
    ports:
      - 2283:2283
    depends_on:
      - redis
      - database
    restart: always
    healthcheck:
      disable: false

  immich-machine-learning:
    container_name: immich_machine_learning
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    volumes:
      - model-cache:/cache
    env_file:
      - .env
    restart: always
    healthcheck:
      disable: false

  redis:
    container_name: immich_redis
    image: docker.io/redis:6.2-alpine@sha256:eaba718fecd1196d88533de7ba49bf903ad33664a92debb24660a922ecd9cac8
    healthcheck:
      test: redis-cli ping || exit 1
    restart: always

  database:
    container_name: immich_postgres
    image: docker.io/tensorchord/pgvecto-rs:pg14-v0.2.0@sha256:90724186f0a3517cf6914295b5ab410db9ce23190a2d9d0b9dd6463e3fa298f0
    volumes:
      - ${DB_DATA_LOCATION}:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
      POSTGRES_INITDB_ARGS: '--data-checksums'
    healthcheck:
      test: >-
        pg_isready --dbname="$${POSTGRES_DB}" --username="$${POSTGRES_USER}" || exit 1;
        Chksum="$$(psql --dbname="$${POSTGRES_DB}" --username="$${POSTGRES_USER}" --tuples-only --no-align
        --command='SELECT COALESCE(SUM(checksum_failures), 0) FROM pg_stat_database')";
        echo "checksum failure count is $$Chksum";
        [ "$$Chksum" = '0' ] || exit 1
      interval: 5m
      start_interval: 30s
      start_period: 5m
    command: >-
      postgres
      -c shared_preload_libraries=vectors.so
      -c 'search_path="$$user", public, vectors'
      -c logging_collector=on
      -c max_wal_size=2GB
      -c shared_buffers=512MB
      -c wal_compression=on
    restart: always

volumes:
  pgdata:
  model-cache:
```

### Your .env content

```Shell
# You can find documentation for all the supported env variables at https://immich.app/docs/install/environment-variables

# The location where your uploaded files are stored
UPLOAD_LOCATION=./library

# The Immich version to use. You can pin this to a specific version like "v1.71.0"
IMMICH_VERSION=release

# Connection secret for postgres. You should change it to a random password
DB_PASSWORD=[CENSORED}


# The values below this line do not need to be changed
###################################################################################
DB_HOSTNAME=immich_postgres
DB_USERNAME=postgres
DB_DATABASE_NAME=immich
DB_DATA_LOCATION=pgdata

REDIS_HOSTNAME=immich_redis
```

### Reproduction steps

1. Open iOS mobile app and upload a photo
2. Check the web interface for the photo. The photo is either showing up with (green) artifacts, indicating that the file is corrupted or no thumbnail is generated (not sure if that is unrelated and fixed by now).
3. Check the iOS app to see the uploaded photo being replicated, one version is the original and one is the corrupted one.

### Relevant log output

```shell
immich_server            | [Nest] 18  - 12/08/2024, 9:28:04 PM     LOG [Api:Bootstrap] Immich Server is listening on http://[::1]:2283 [v1.122.1] [production] 
immich_server            | [Nest] 18  - 12/08/2024, 9:28:17 PM     LOG [Api:EventRepository] Websocket Connect:    M_FX6JytsMXxu2DzAAAB
immich_server            | [Nest] 18  - 12/08/2024, 9:28:19 PM     LOG [Api:EventRepository] Websocket Disconnect: M_FX6JytsMXxu2DzAAAB
immich_server            | [Nest] 18  - 12/08/2024, 9:28:20 PM     LOG [Api:EventRepository] Websocket Connect:    4QtHIi_eTn3YIYZ1AAAD
immich_server            | [Nest] 18  - 12/08/2024, 9:28:30 PM     LOG [Api:EventRepository] Websocket Connect:    mPXfSQvIA_W4POYGAAAF       
immich_server            | [Nest] 18  - 12/08/2024, 9:28:32 PM     LOG [Api:EventRepository] Websocket Connect:    9H78SuAWjaXxJ6-YAAAH
immich_machine_learning  | [12/08/24 21:28:39] INFO     Loading detection model 'buffalo_l' to memory                                         
immich_machine_learning  | [12/08/24 21:28:39] INFO     Setting execution providers to                     
immich_machine_learning  |                              ['CPUExecutionProvider'], in descending order of                       
immich_machine_learning  |                              preference                                          
immich_machine_learning  | [12/08/24 21:28:39] INFO     Loading visual model 'ViT-B-32__openai' to memory  
immich_machine_learning  | [12/08/24 21:28:39] INFO     Setting execution providers to                     
immich_machine_learning  |                              ['CPUExecutionProvider'], in descending order of   
immich_machine_learning  |                              preference                                          
immich_machine_learning  | [12/08/24 21:28:42] INFO     Loading recognition model 'buffalo_l' to memory    
immich_machine_learning  | [12/08/24 21:28:42] INFO     Setting execution providers to                     
immich_machine_learning  |                              ['CPUExecutionProvider'], in descending order of   
immich_machine_learning  |                              preference                                          
immich_server            | [Nest] 18  - 12/08/2024, 9:28:59 PM     LOG [Api:EventRepository] Websocket Disconnect: 9H78SuAWjaXxJ6-YAAAH
immich_server            | [Nest] 18  - 12/08/2024, 9:29:22 PM     LOG [Api:EventRepository] Websocket Disconnect: 4QtHIi_eTn3YIYZ1AAAD
immich_server            | [Nest] 18  - 12/08/2024, 9:29:23 PM     LOG [Api:EventRepository] Websocket Connect:    4fqbhA9d9TETaplXAAAJ
immich_server            | [Nest] 18  - 12/08/2024, 9:29:40 PM     LOG [Api:EventRepository] Websocket Disconnect: mPXfSQvIA_W4POYGAAAF
```

**Comment by alextran1502** [comment]
Also, please try to bring your instance down, remove the images, repull and then bring the instance up and try uploading again

**Comment by obale** [comment]
I removed in the past multiple times the containers (via `docker compose down`), now I also removed all the images via `docker rmi -f $(docker images -aq)`.  For good measure I was also running `docker system prune`.  After starting immich via `docker compose up -d` . Same problem, no luck.

The photos have some sensitive meta data in them, such as location data, but I will see what I can do. In the meantime the screenshot below shows the artifact issue. 

![Screenshot From 2024-12-09 12-51-59](https://github.com/user-attachments/assets/e2ea6d03-a304-4ffe-bbb4-0bdcae34da75)

Some of the maybe relevant meta data is:

```
IMG_1298.HEIC
upload/library/obale/2024/2024-12-08/IMG_1298+2.HEIC

9 MP 2268 x 4032 
1.2 MiB

Apple iPhone 11 Pro
ƒ/1.8 1/60 s 4.25 mm ISO 200
```

I have a pretty standard install on an Ubuntu VM in Proxmox, so I hoped someone had experienced the same issue.

**Comment by obale** [comment]
I shot a completely new photo and remove the location from the iOS Camera app (send this you, @alextran1502, via email). Some interesting observations:

* It corrupts the image after upload, as outlined in this ticket.
* It adds back the location to the meta data, that is why I shot a completely new photo to assure there is no overlap with the file name or image content hash. Potentially it derives the information from other meta data fields.

Hope that helps.

**Comment by obale** [comment]
I am wondering if I can help debug this issue better by understanding when and how the uploaded photos are modified. Can someone point me to the right direction to look into that or provide some context how the original files are modified after or during successful uploads?  Thank you!

**Comment by obale** [comment]
Not conclusive, because of small sample size, but it looks like the images corrupted are from the local Camera (latest iOS and immich app version). Photos shared seem to upload just fine.

**Comment by alextran1502** [comment]
Uploaded the file. This is how it looks from my instance.

<img width="1322" alt="image" src="https://github.com/user-attachments/assets/d8a8aaba-ee88-4b3e-b70d-d61980326ce6" />

It doesn't look corrupted from the screenshot, correct?

What is your camera settings?

**Comment by alextran1502** [comment]
Image is not modified during upload 🤔

**Comment by alextran1502** [comment]
Do you mind try turn the phone off and on. Then try uploading a newly taken photo?

**Comment by obale** [comment]
That is strange that they are not modified, but somehow corrupted consistently and randomly. I restarted my phone multiple times, same result. I even removed the immich app and installed it again.

**Comment by alextran1502** [comment]
Perhaps you can try creating a new Immich instance with brand new docker images and try uploading photo to that instance

**Comment by akostadinov** [comment]
@alextran1502 , there might be some issue with upload handling because I see corruption uploading via web. See #14657

There is a link to a test file. It might be related to server being underpowered or something. Given client is doing the hashing, I wonder if server checks the hash after file being uploaded.

**Comment by obale** [comment]
Seems to be corrupted during upload, because when I am downloading via ssh (scp) and are opening local I see the same artifacts. Didn't compare the hash against the original, given that it is obviously corrupted and it can be visually verified.

Trying now to upgrade to the latest version published today.

**Comment by akostadinov** [comment]
@obale then also good to check whether images downloaded from phone directly open properly. Maybe the format is not good for your PC?

**Comment by alextran1502** [comment]
@obale Are you using iCloud-optimized storage?

**Comment by obale** [comment]
The problem may be the AAE format. It stores the images in JPG and AAE format. As far as I can tell it uploads AAE. I am checking now if I am using iCloud optimized storage, but I am not backing up the photos to iCloud.

**Comment by alextran1502** [comment]
How do you set your phone to capture the photo in those formats?

**Comment by obale** [comment]
Even in JPG format they are corrupted, not always, but most times.

**Comment by obale** [comment]
Just as reference and because resources have been mentioned. I am running immich in a Proxmox VM (Ubuntu 24.04.1 LTS) with 24 GB of memory and 15 processors (2 Sockets and 8 Cores). Nothing else is running in the VM.

**Comment by alextran1502** [comment]
What was is the format you have selected? Can you share the screenshot?

**Comment by obale** [comment]
I tried both formats.

**Comment by alextran1502** [comment]
Thanks. I can't reproduce this issue with all format. We will need to be able to reproduce this for a fix.

Can you try create a new test Immich instance with all stock settings and try uploading some photos there?

**Comment by obale** [comment]
Will do that.

**Comment by obale** [comment]
It took some time, I wanted to make sure server-side everything is setup correctly. As recommended I setup a completely new instance and was restoring from a backup that I knew is working (restic backup that also checks the hashes). 

I re-run all the jobs and from the web interface it looks all good. 

**Same problem with uploads from the iOS app**. I even completely removed it during the new setup and installed it after the remote instance was ready, i.e., everything running and all the jobs finished.

Also, the underlying host is using ZFS and I checked for file system and hard disk corruption. No issues there. I am running out of ideas.

**Comment by alextran1502** [comment]
Can you create a test new instance, with new account and everything? If it still happens then I will resort to one of the strange case that cannot be reproduced. 

So the next step is to test on a different host

**Comment by akostadinov** [comment]
@obale , if you send a PGP public key, I can setup a temporary account on my instance for you so that you can try if it still corrupts there.

**Comment by obale** [comment]
Fresh new installed with no data and no database has the same issue. 

My PGP key can be found here: https://github.com/obale.gpg

**Comment by obale** [comment]
@akostadinov, I received the immich instance information and tried it out. With this instance it works without corrupting the images. Not sure what is different on my machines. Will continue debugging from my side. Really appreciate your help  in this matter.

Please feel free to delete the temporary account and shutdown the instance again. Thank you!

**Comment by obale** [comment]
Created a completely new instance on a physically different machine (TrueNAS Scale, also Ubuntu based). Same issue with corrupted files. I enabled the storage template there too, with the default values. So it is either the storage template or the HAProxy reverse proxy with TLS termination that is before the machines.

**Comment by obale** [comment]
From the logs (new instance on TrueNAS).

```
immich_server            | [Nest] 17  - 12/19/2024, 6:38:17 PM     LOG [Api:EventRepository] Websocket Connect:    -ql0OcP2fr2e5MurAAAN
immich_server            | [Nest] 7  - 12/19/2024, 6:38:30 PM   ERROR [Microservices:JobService] Unable to run job handler (thumbnailGeneration/generate-thumbnails): Error: VipsJpeg: ./lib/jpegli/decode_scan.cc:539: Failed to decode DCT block
immich_server            | [Nest] 7  - 12/19/2024, 6:38:30 PM   ERROR [Microservices:JobService] Error: VipsJpeg: ./lib/jpegli/decode_scan.cc:539: Failed to decode DCT block
immich_server            |     at Sharp.toBuffer (/usr/src/app/node_modules/sharp/lib/output.js:163:17)
immich_server            |     at MediaRepository.decodeImage (/usr/src/app/dist/repositories/media.repository.js:57:68)
immich_server            |     at MediaService.generateImageThumbnails (/usr/src/app/dist/services/media.service.js:161:63)
immich_server            |     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
immich_server            |     at async MediaService.handleGenerateThumbnails (/usr/src/app/dist/services/media.service.js:111:25)
immich_server            |     at async JobService.onJobStart (/usr/src/app/dist/services/job.service.js:148:28)
immich_server            |     at async EventRepository.onEvent (/usr/src/app/dist/repositories/event.repository.js:134:13)
immich_server            |     at async Worker.processJob (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:394:28)
immich_server            |     at async Worker.retryIfFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:581:24)
immich_server            | [Nest] 7  - 12/19/2024, 6:38:30 PM   ERROR [Microservices:JobService] Object:
immich_server            | {
immich_server            |   "id": "198e32fe-4303-473b-b72a-e6195d2dd6da",
immich_server            |   "source": "upload"
immich_server            | }
immich_server            | 
immich_server            | [Nest] 17  - 12/19/2024, 6:38:31 PM     LOG [Api:EventRepository] Websocket Disconnect: -ql0OcP2fr2e5MurAAAN
immich_server            | [Nest] 17  - 12/19/2024, 6:38:32 PM     LOG [Api:EventRepository] Websocket Connect:    xn87tCgfueHALfgiAAAP
```

**Comment by obale** [comment]
I am closing this issue. Thank you for your patience to help debug the issue. 

Not 100% sure what caused it, but it wasn't an Immich issue, at least not alone. After moving the reverse proxy from pfSense (firewall) haproxy to traefik via docker compose (same machine as immich) it seems to work as expected. 

**tldr;** It is working now after changing the reverse proxy.

**Comment by akostadinov** [comment]
Yeah, it is not the template because I also have it enabled. Sounds like the issue was with your reverse proxy.

I'm worried though, imimch shouldn't allow silently corrupted uploads regardless of network and proxy issues.

The client generated a hash so server should verify that the upload has that expected hash. Same with downloads. The client should locally validate the hash when possible.

So in my opinion, although this problem has been resolved with changing the reverse proxy, I think there is still an issue to be improved in the immich codebase.

**Comment by obale** [comment]
@akostadinov, I completely agree. It is also strange that my reverse proxy configuration didn't change and is relatively simple, i.e., nothing exotic. No matter how I spin it I can't explain the issues. I am also running paperless-ngx behind the same reverse proxy configuration and it seems to work perfectly fine, at least for now. To be fair immich was also working until 2 weeks ago.

**Comment by jacobmee** [comment]
Hey, guys
I met a smilar issue under [v1.122.2] - iPhone creates HEIC file which can't generate thumbfile which shows broken in web/ios. and I made change to let IOS only generate compatible JPG version, which solved the problem.

I also did a research seem you guy already fix the "libheif" issue in https://github.com/immich-app/immich/issues/10464.

However, I did multiple round of test, so pretty much sure it's the thumbfile can't be generated.  Also attached the logs:

Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:   "source": "upload"
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:   "id": "8bbeef33-81bc-4abd-a776-87ab2876ba1b",
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]: {
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]: [Nest] 121392  - 12/21/2024, 3:31:31 PM   ERROR [Microservices:JobService] >
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at async Worker.retryIfFailed (/snap/immich-distribution/170/usr/src/ap>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at async Worker.processJob (/snap/immich-distribution/170/usr/src/app/n>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at async EventRepository.onEvent (/snap/immich-distribution/170/usr/src>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at async JobService.onJobStart (/snap/immich-distribution/170/usr/src/a>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at async MediaService.handleGenerateThumbnails (/snap/immich-distributi>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at process.processTicksAndRejections (node:internal/process/task_queues>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at MediaService.generateImageThumbnails (/snap/immich-distribution/170/>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at MediaRepository.decodeImage (/snap/immich-distribution/170/usr/src/a>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]:     at Sharp.toBuffer (/snap/immich-distribution/170/usr/src/app/node_modul>
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]: heif: Invalid input: Unspecified: Too many auxiliary image references (2.0)
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]: [Nest] 121392  - 12/21/2024, 3:31:31 PM   ERROR [Microservices:JobService] >
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]: heif: Invalid input: Unspecified: Too many auxiliary image references (2.0)
Dec 21 15:31:31 HOME immich-distribution.immich-server[121392]: [Nest] 121392  - 12/21/2024, 3:31:31 PM   ERROR [Microservices:JobService] >

<img width="467" alt="image" src="https://github.com/user-attachments/assets/01c4a188-6177-480e-aa45-37fa55407c26" />

**Comment by alextran1502** [comment]
@jacobmee you are using the Snap distribution which is not officially supported by us, it is probably packaged with different version of the dependencies version we are using, thus the issue.

Please use the official docker-compose method instead

**Comment by jacobmee** [comment]
Thanks Alex!
Yep, I understand the difference and it works well for more than 6 months and I don’t want to migrate which obviously will cost me considerable effort since I have over 40k photos there already.

Th current workaround for me is 1) don’t use HEIC 2) Converting to jpg myself before uploading/copy to external lib.  Both works fine.

It’s fine and thanks again if it only happens to snap version.

**Comment by alextran1502** [comment]
@jacobmee Yeah, unfortunately, that should be the issue you bring to that package maintainer
